pkg/plugins/pg/destination/destination.go

```diff
@@ -16,114 +16,178 @@ package destination
 
 import (
 	"context"
-	"database/sql"
+	"encoding/json"
 	"fmt"
 	"strings"
 
-	"github.com/conduitio/conduit/pkg/foundation/cerrors"
-	"github.com/conduitio/conduit/pkg/plugins"
-	"github.com/conduitio/conduit/pkg/record"
+	"github.com/conduitio/conduit/pkg/plugin/sdk"
 
 	sq "github.com/Masterminds/squirrel"
-	_ "github.com/lib/pq" // Blank import to register the PostgreSQL driver
+	"github.com/jackc/pgx/v4"
 )
 
-// Destination fulfills the Destination interface
-type Destination struct {
-	Position record.Position
+// Postgres requires use of a different variable placeholder.
+var psql = sq.StatementBuilder.PlaceholderFormat(sq.Dollar)
 
-	db *sql.DB
+type Adapter struct {
+	sdk.UnimplementedDestination
+
+	Position sdk.Position
+
+	conn   *pgx.Conn
+	config config
 }
 
-// Open attempts to open a database connection to Postgres.
-func (s *Destination) Open(ctx context.Context, cfg plugins.Config) error {
-	url, ok := cfg.Settings["url"]
-	if !ok {
-		return cerrors.New("must provide a connection URL")
+const (
+	actionDelete = "delete"
+	actionInsert = "insert"
+	actionUpdate = "update"
+)
+
+type config struct {
+	url           string
+	tableName     string
+	keyColumnName string
+}
+
+func New() sdk.Destination {
+	return &Adapter{}
+}
+
+func (d *Adapter) Configure(ctx context.Context, cfg map[string]string) error {
+	d.config = config{
+		url:           cfg["url"],
+		tableName:     cfg["table"],
+		keyColumnName: cfg["keyColumnName"],
 	}
-	db, err := sql.Open("postgres", url)
-	if err != nil {
-		return cerrors.Errorf("failed to open source DB: %w", err)
+	return nil
+}
+
+func (d *Adapter) Open(ctx context.Context) error {
+	if err := d.connect(ctx, d.config.url); err != nil {
+		return fmt.Errorf("failed to connecto to postgres: %w", err)
 	}
-	err = db.Ping()
+	return nil
+}
+
+func (d *Adapter) Write(ctx context.Context, record sdk.Record) error {
+	return d.write(ctx, record)
+}
+
+func (d *Adapter) Flush(context.Context) error {
+	return nil
+}
+
+func (d *Adapter) Teardown(ctx context.Context) error {
+	return d.conn.Close(ctx)
```

Comment:
We need to check if `d.conn` is even set, otherwise this produces a panic on connector creation, because all that gets called is `Configure` and then `Teardown`.